### PR TITLE
Ex3

### PR DESCRIPTION
**Pull request by orimaromcom**

### CODE REVIEW

**Comment by orimaromcom** [comment]
Hi All,
This ex3 is ready.
You can read all about it in the README file, watch the demo, or dive into the code.
Enjoy :)

https://user-images.githubusercontent.com/66553796/172823555-556de2ea-7291-43b7-aa53-7734ed8d017c.mov

File: src/ex3/services/itemManager.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import PokemonClient from "./pokemonClient.js";
+import Item from "../models/item.js";
+import * as config from "../conf/conf.js";
+import fs from "node:fs";
+import chalk from "chalk";
+
+const ID_INIT_KEY = 2000;
+class ItemManager {
```
Comment (edogr1) [inline src/ex3/services/itemManager.js:8]: This class serves 2 purposes and could be seperated. currently it handles both the list of items (add, remove, pokemon related stuff), and handling the storage itself.
The "load" and "save" methods could be separated to a different class of file storage.

File: src/ex3/services/itemManager.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+   load() {
+    try {
+      if (!fs.existsSync(config.DB_PATH_FILENAME)) {
+        this.createStorageFile(
+          config.DB_PATH_DIRECTORY,
+          config.DB_PATH_FILENAME
+        );
+      }
+      this.itemList =  JSON.parse(fs.readFileSync(config.DB_PATH_FILENAME));
+      this.idGenerator = ID_INIT_KEY + this.itemList.length;
```
Comment (edogr1) [inline src/ex3/services/itemManager.js:24]: adding a unique ID is a good idea but this one is buggy, if I'll delete an item, and next item will get the same ID as the one before. You can use GUID or UUID instead

File: src/ex3/services/pokemonClient.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      }
+      return errorResponse;
+    } catch (error) {
+      console.log("this is error");
+      return errorResponse;
+    }
+  }
+
+  fetchMultiplePokemons(pokemonsToFetch) {
+    const promises = [];
```
Comment (edogr1) [inline src/ex3/services/pokemonClient.js:31]: return pokemonsToFetch.map(this.fetchPokemon);

File: src/ex3/services/parser.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  "ivysaur",
+  "venusaur",
+  "charmander",
+  "pikachu",
+];
+
+// Implement the `Main` class here
+class Parser {
+  constructor() {
+    this.itemManager = new ItemManager();
```
Comment (edogr1) [inline src/ex3/services/parser.js:15]: you're not using the itemManager class here.

File: src/ex3/services/itemManager.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    try {
+      fs.writeFileSync(config.DB_PATH_FILENAME, JSON.stringify(this.itemList));
+    } catch (error) {
+      console.error("cannot save", error);
+    }
+  }
+
+  addMultipleTasksToList(tasks) {
+    for (let task of tasks) {
+      this.addToItemList(task, false);
```
Comment (edogr1) [inline src/ex3/services/itemManager.js:49]: By doing this you're saving the file after each task, instead of adding all the tasks to the array and save only once.

File: src/ex3/services/itemManager.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        this.idGenerator,
+        isPokemon,
+        pokemonData.sprites.front_default
+      );
+    } else {
+      this.item = new Item(itemTextValue, this.idGenerator, isPokemon);
+    }
+    this.idGenerator++;
+    this.itemList.push(this.item);
+     await this.save();
```
Comment (edogr1) [inline src/ex3/services/itemManager.js:68]: the saving of the list shouldn't be done from here. it should be called from the commands when the process is done. (see above comment about addMultipleTasksToList)

File: src/ex3/commands/addCommand.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  const pokemonClient = new PokemonClient();
+  await itemManager.load()
+
+  const dictionary = parser.parseInputValue(inputValue);
+  itemManager.addMultipleTasksToList(dictionary.tasks);
+  const fetchedPokemons = pokemonClient.fetchMultiplePokemons(
+    dictionary.pokemons
+  );
+
+  Promise.all(fetchedPokemons).then((values) => {
```
Comment (edogr1) [inline src/ex3/commands/addCommand.js:18]: all of the logic that is not related to print the result should be inside of the ItemManager class. This class should parse the input, call the itemManager (load, add, save) and print a response.

File: src/ex3/commands/showCommand.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  itemList.forEach((item) => {
+    let toString = itemList.indexOf(item) + 1;
+    toString += ". ";
+
+    if (item.isPokemon) {
+      toString += `Catch ${item.name} `;
+      console.log(chalk.cyan(toString));
+      asciifyImage(
```
Comment (edogr1) [inline src/ex3/commands/showCommand.js:32]: if you would use async/await instead of passing a call you could print the image directly after the task instead of printing them all at the end.

File: src/ex3/commands/addCommand.js (hunk 7 above)
Comment (edogr1) [reply src/ex3/commands/addCommand.js:18]: also using the pokemon client should be done inside of the item manager

File: src/ex3/commands/deleteAllCommand.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import chalk from "chalk";
+import Parser from "../services/parser.js";
+import ItemManager from "../services/itemManager.js";
+
+export async function deleteAll() {
+  const itemManager = new ItemManager();
+  
+  itemManager.load();
+  itemManager.itemList = []
```
Comment (edogr1) [inline src/ex3/commands/deleteAllCommand.js:9]: In general it's better add a dedicated method of "deleteAll" in the itemMamager.
If you would like to save in a List instead of an array for example, you'll have to change it both here and in the itemManager.

File: src/ex3/commands/showCommand.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+import chalk from "chalk";
+import chalkAnimation from "chalk-animation";
+import Parser from "../services/parser.js";
+import ItemManager from "../services/itemManager.js";
+import * as config from "../conf/conf.js";
+import fs from "node:fs";
+import asciifyImage from "asciify-image";
+
+const sleep = (ms = 2000) => new Promise((r) => setTimeout(r, ms));
```
Comment (edogr1) [inline src/ex3/commands/showCommand.js:9]: not that you're using this in 2 places, this could be exported to a utils class.

File: src/ex3/commands/addCommand.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+      const isPokemonExist = itemManager.isPokemonExists(pokemonName);
+
+      if (isPokemonExist) {
+        console.log(chalk.red(`${pokemonName} is already in list`));
+        continue; //if pokemon exist continue to the next pokemon
+      }
+
+      if (!value.error && !isPokemonExist) {
+        // pokemon does not exist and is not error , add it to the list
+        itemManager.addToItemList(pokemonName, true, pokemonData);
```
Comment (edogr1) [inline src/ex3/commands/addCommand.js:34]: note that also here the file will be saved after each pokemon is added.

File: src/ex3/commands/inquirerCommand.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  itemManager.load();
+  const itemList = itemManager.itemList;
+
+  inquirer
+    .prompt([
+      {
+        type: "list",
+        message: "\n\nPick a task to remove",
+        name: "action",
+        choices: ["add task", "delete task", "delete list", "get list"],
```
Comment (edogr1) [inline src/ex3/commands/inquirerCommand.js:20]: you can use consts for the choices, that way if you will want to change the text you'll have to change it only in 1 place.

**Review by edogr1 — COMMENTED** [review]
Very good submission!
Well done for implementing all the features including the bonuses!
Overall the code was really readable and understandable, good thinking about how to divide classes by responsibility.
My main comment is about separating completely the way you're accessing the storage (file in this case), and paying attention to saving only when you need and not after every change.
Keep it going with the good work!
